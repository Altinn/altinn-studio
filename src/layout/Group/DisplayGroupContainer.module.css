.groupTitle {
  font-weight: 500;
  font-size: 1.25rem;
  margin-bottom: 0;
}

.groupBody {
  margin-top: 12px;
  margin-bottom: 0;
}

.groupContainer {
  padding-bottom: 38px !important;
}

.groupingIndicator {
  border-left: 2px solid #949494;
  margin-left: 12px !important;
  margin-bottom: 12px !important;
  margin-top: 12px !important;
  padding-top: 0 !important;
  padding-bottom: 0 !important;
}
