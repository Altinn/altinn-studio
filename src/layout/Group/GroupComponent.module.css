.groupWrapper {
  padding-top: var(--fds-spacing-3);
  padding-bottom: var(--fds-spacing-9);
  width: 100%;

  .legend,
  .description {
    padding: 0 var(--fds-spacing-3);
  }
}

.panelWrapper {
  padding: var(--fds-spacing-3);
  padding-bottom: 0;
  width: 100%;

  &.summary {
    padding: 0;
  }
}

.groupContainer {
  display: flex;
  flex-wrap: wrap;

  > * {
    padding: var(--fds-spacing-3);
  }
}

.indented {
  border-left: 2px solid #949494;
  padding: var(--fds-spacing-3);
  padding-right: 0;
  margin-left: var(--fds-spacing-3);
}

.summary {
  width: 100%;

  .groupContainer > * {
    padding: var(--fds-spacing-3) 0;
  }

  .legend,
  .description {
    padding: 0;
  }

  .indented {
    margin-left: 0;
  }
}
