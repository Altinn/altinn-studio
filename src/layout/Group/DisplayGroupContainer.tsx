import React from 'react';

import { Heading } from '@digdir/design-system-react';
import { Grid } from '@material-ui/core';
import cn from 'classnames';

import { ConditionalWrapper } from 'src/components/ConditionalWrapper';
import { useLanguage } from 'src/hooks/useLanguage';
import classes from 'src/layout/Group/DisplayGroupContainer.module.css';
import { pageBreakStyles } from 'src/utils/formComponentUtils';
import { LayoutNode } from 'src/utils/layout/LayoutNode';
import type { HeadingLevel } from 'src/types/shared';
import type { LayoutNodeFromType } from 'src/utils/layout/hierarchy.types';

export interface IDisplayGroupContainer {
  groupNode: LayoutNodeFromType<'Group'>;
  id?: string;
  onlyRowIndex?: number | undefined;
  renderLayoutNode: (node: LayoutNode) => JSX.Element | null;
}

const headingSizes: { [k in HeadingLevel]: Parameters<typeof Heading>[0]['size'] } = {
  [2]: 'medium',
  [3]: 'small',
  [4]: 'xsmall',
  [5]: 'xsmall',
  [6]: 'xsmall',
};

export function DisplayGroupContainer({ groupNode, id, onlyRowIndex, renderLayoutNode }: IDisplayGroupContainer) {
  const { lang, langAsString } = useLanguage();
  const container = groupNode.item;
  const title = langAsString(container.textResourceBindings?.title);
  const body = lang(container.textResourceBindings?.body);

  if (groupNode.isHidden()) {
    return null;
  }

  const isNested = groupNode.parent instanceof LayoutNode;
  const headingLevel = Math.min(Math.max(groupNode.parents().length + 1, 2), 6) as HeadingLevel;
  const headingSize = headingSizes[headingLevel];

  return (
    <Grid
      container={true}
      item={true}
      id={id || container.id}
      className={cn(pageBreakStyles(container.pageBreak), {
        [classes.groupContainer]: !isNested,
        [classes.groupingIndicator]: !!container.showGroupingIndicator && isNested,
      })}
      spacing={3}
      alignItems='flex-start'
      data-testid='display-group-container'
      data-componentid={container.id}
    >
      {(title || body) && (
        <Grid
          item={true}
          xs={12}
        >
          {title && (
            <Heading
              level={headingLevel}
              size={headingSize}
            >
              {title}
            </Heading>
          )}
          {body && <p className={classes.groupBody}>{body}</p>}
        </Grid>
      )}
      <ConditionalWrapper
        condition={!!container.showGroupingIndicator && !isNested}
        wrapper={(children) => (
          <Grid
            item={true}
            container={true}
            spacing={3}
            alignItems='flex-start'
            className={classes.groupingIndicator}
          >
            {children}
          </Grid>
        )}
      >
        <>{groupNode.children(undefined, onlyRowIndex).map((n) => renderLayoutNode(n))}</>
      </ConditionalWrapper>
    </Grid>
  );
}
