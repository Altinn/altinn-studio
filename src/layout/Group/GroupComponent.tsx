import React from 'react';
import type { JSX } from 'react';

import { Heading } from '@digdir/designsystemet-react';
import cn from 'classnames';

import { Fieldset } from 'src/app-components/Label/Fieldset';
import { Panel } from 'src/app-components/Panel/Panel';
import { ConditionalWrapper } from 'src/components/ConditionalWrapper';
import { FullWidthWrapper } from 'src/components/form/FullWidthWrapper';
import { Lang } from 'src/features/language/Lang';
import classes from 'src/layout/Group/GroupComponent.module.css';
import { BaseLayoutNode } from 'src/utils/layout/LayoutNode';
import { Hidden } from 'src/utils/layout/NodesContext';
import { useNodeDirectChildren, useNodeItem } from 'src/utils/layout/useNodeItem';
import { useNodeTraversal } from 'src/utils/layout/useNodeTraversal';
import type { HeadingLevel } from 'src/layout/common.generated';
import type { LayoutNode } from 'src/utils/layout/LayoutNode';
import type { TraversalRestriction } from 'src/utils/layout/useNodeTraversal';

export interface IGroupComponent {
  groupNode: LayoutNode<'Group'>;
  containerDivRef?: React.Ref<HTMLDivElement>;
  id?: string;
  restriction?: TraversalRestriction;
  isSummary?: boolean;
  renderLayoutNode: (node: LayoutNode) => JSX.Element | null;
}

const headingSizes: { [k in HeadingLevel]: Parameters<typeof Heading>[0]['size'] } = {
  [2]: 'medium',
  [3]: 'small',
  [4]: 'xsmall',
  [5]: 'xsmall',
  [6]: 'xsmall',
};

export function GroupComponent({
  groupNode,
  containerDivRef,
  id,
  restriction,
  isSummary,
  renderLayoutNode,
}: IGroupComponent) {
  const container = useNodeItem(groupNode);
  const { title, summaryTitle, description } = container.textResourceBindings ?? {};
  const isHidden = Hidden.useIsHidden(groupNode);

  const children = useNodeDirectChildren(groupNode, restriction);
  const depth = useNodeTraversal((traverser) => traverser.with(groupNode).parents().length);

  if (isHidden) {
    return null;
  }

  const isNested = groupNode.parent instanceof BaseLayoutNode;
  const isPanel = container.groupingIndicator === 'panel';
  const isIndented = container.groupingIndicator === 'indented';
  const headingLevel = container.headingLevel ?? (Math.min(Math.max(depth + 1, 2), 6) as HeadingLevel);
  const headingSize = headingSizes[headingLevel];
  const legend = isSummary ? (summaryTitle ?? title) : title;

  return (
    <div className={cn(classes.groupWrapper, { [classes.panelWrapper]: isPanel, [classes.summary]: isSummary })}>
      <ConditionalWrapper
        condition={isPanel && !isSummary}
        wrapper={(child) => (
          <FullWidthWrapper>
            <Panel variant='info'>{child}</Panel>
          </FullWidthWrapper>
        )}
      >
        <Fieldset
          legend={
            legend ? (
              <Heading
                className={classes.legend}
                level={headingLevel}
                size={headingSize}
              >
                <Lang id={legend} />
              </Heading>
            ) : undefined
          }
          description={
            description && !isSummary ? (
              <div className={classes.description}>
                <Lang id={description} />
              </div>
            ) : undefined
          }
        >
          <div
            data-componentid={container.id}
            data-componentbaseid={container.baseComponentId || container.id}
            ref={containerDivRef}
            id={id ?? container.id}
            data-testid='display-group-container'
            className={cn(classes.groupContainer, {
              [classes.indented]: isIndented && !isNested,
            })}
          >
            {children.map((n) => renderLayoutNode(n))}
          </div>
        </Fieldset>
      </ConditionalWrapper>
    </div>
  );
}
