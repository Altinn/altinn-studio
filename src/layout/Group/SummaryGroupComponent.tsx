import React, { useCallback } from 'react';

import { ErrorPaper } from 'src/components/message/ErrorPaper';
import { Lang } from 'src/features/language/Lang';
import { useLanguage } from 'src/features/language/useLanguage';
import { useDeepValidationsForNode } from 'src/features/validation/selectors/deepValidationsForNode';
import { hasValidationErrors } from 'src/features/validation/utils';
import { CompCategory } from 'src/layout/common';
import { GroupComponent } from 'src/layout/Group/GroupComponent';
import classes from 'src/layout/Group/SummaryGroupComponent.module.css';
import { EditButton } from 'src/layout/Summary/EditButton';
import { SummaryComponent } from 'src/layout/Summary/SummaryComponent';
import { Hidden } from 'src/utils/layout/NodesContext';
import { useNodeDirectChildren, useNodeItem } from 'src/utils/layout/useNodeItem';
import type { CompTypes, ITextResourceBindings } from 'src/layout/layout';
import type { SummaryRendererProps } from 'src/layout/LayoutComponent';
import type { LayoutNode } from 'src/utils/layout/LayoutNode';

export function SummaryGroupComponent({
  onChangeClick,
  changeText,
  summaryNode,
  targetNode,
  overrides,
}: SummaryRendererProps<'Group'>) {
  const summaryItem = useNodeItem(summaryNode);
  const targetItem = useNodeItem(targetNode);
  const excludedChildren = summaryItem?.excludedChildren;
  const display = overrides?.display || summaryItem?.display;
  const { langAsString } = useLanguage();
  const isHidden = Hidden.useIsHiddenSelector();

  const inExcludedChildren = useCallback(
    (n: LayoutNode) =>
      excludedChildren ? excludedChildren.includes(n.id) || excludedChildren.includes(n.baseId) : false,
    [excludedChildren],
  );

  const groupValidations = useDeepValidationsForNode(targetNode);
  const groupHasErrors = hasValidationErrors(groupValidations);

  const textBindings = targetItem.textResourceBindings as ITextResourceBindings;
  const summaryAccessibleTitleTrb =
    textBindings && 'summaryAccessibleTitle' in textBindings ? textBindings.summaryAccessibleTitle : undefined;
  const summaryTitleTrb = textBindings && 'summaryTitle' in textBindings ? textBindings.summaryTitle : undefined;
  const titleTrb = textBindings && 'title' in textBindings ? textBindings.title : undefined;
  const ariaLabel = langAsString(summaryAccessibleTitleTrb ?? summaryTitleTrb ?? titleTrb);
  const children = useNodeDirectChildren(targetNode).filter((n) => !inExcludedChildren(n));

  const largeGroup = overrides?.largeGroup ?? summaryItem?.largeGroup ?? false;
  if (largeGroup) {
    return (
      <>
        {
          <GroupComponent
            key={`summary-${targetNode.id}`}
            id={`summary-${targetNode.id}`}
            groupNode={targetNode}
            isSummary={true}
            renderLayoutNode={(node) => (
              <SummaryComponentFromNode
                targetNode={node}
                summaryNode={summaryNode}
                overrides={overrides}
                inExcludedChildren={inExcludedChildren}
              />
            )}
          />
        }
      </>
    );
  }

  const childSummaryComponents = children.map((child) => {
    if (!child.isCategory(CompCategory.Form) || isHidden(child)) {
      return;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const RenderCompactSummary = child.def.renderCompactSummary.bind(child.def) as React.FC<SummaryRendererProps<any>>;
    return (
      <RenderCompactSummary
        onChangeClick={onChangeClick}
        changeText={changeText}
        key={child.id}
        targetNode={child}
        summaryNode={summaryNode}
        overrides={{}}
      />
    );
  });

  return (
    <>
      <div
        data-testid={'summary-group-component'}
        style={{ width: '100%' }}
      >
        <div className={classes.container}>
          <span className={classes.label}>
            <Lang id={summaryTitleTrb ?? titleTrb} />
          </span>

          {!display?.hideChangeButton && (
            <EditButton
              onClick={onChangeClick}
              editText={changeText}
              label={ariaLabel}
            />
          )}
        </div>
        <div style={{ width: '100%' }}>
          <div className={classes.border}>{childSummaryComponents}</div>
        </div>
      </div>

      {groupHasErrors && !display?.hideValidationMessages && (
        <div className={classes.gridStyle}>
          <ErrorPaper message={langAsString('group.row_error')} />
          <div>
            {!display?.hideChangeButton && (
              <button
                className={classes.link}
                onClick={onChangeClick}
                type='button'
              >
                <Lang id={'form_filler.summary_go_to_correct_page'} />
              </button>
            )}
          </div>
        </div>
      )}
    </>
  );
}

interface SummaryComponentFromRefProps
  extends Pick<SummaryRendererProps<CompTypes>, 'targetNode' | 'summaryNode' | 'overrides'> {
  inExcludedChildren: (node: LayoutNode) => boolean;
}

function SummaryComponentFromNode({
  targetNode,
  inExcludedChildren,
  summaryNode,
  overrides,
}: SummaryComponentFromRefProps) {
  const isHidden = Hidden.useIsHidden(targetNode);
  if (inExcludedChildren(targetNode) || isHidden) {
    return null;
  }

  return (
    <SummaryComponent
      summaryNode={summaryNode}
      overrides={{
        ...overrides,
        targetNode,
        grid: {},
        largeGroup: true,
      }}
    />
  );
}
