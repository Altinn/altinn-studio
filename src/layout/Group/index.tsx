import React from 'react';
import type { JSX } from 'react';

import type { ErrorObject } from 'ajv';

import { GroupDef } from 'src/layout/Group/config.def.generated';
import { GroupRenderer } from 'src/layout/Group/GroupRenderer';
import { GroupHierarchyGenerator } from 'src/layout/Group/hierarchy';
import { LayoutNodeForGroup } from 'src/layout/Group/LayoutNodeForGroup';
import { SummaryGroupComponent } from 'src/layout/Group/SummaryGroupComponent';
import {
  groupIsNonRepeatingExt,
  groupIsNonRepeatingPanelExt,
  groupIsRepeatingExt,
  groupIsRepeatingLikertExt,
} from 'src/layout/Group/tools';
import { runValidationOnNodes } from 'src/utils/validation/validation';
import { buildValidationObject } from 'src/utils/validation/validationHelpers';
import type { LayoutValidationCtx } from 'src/features/devtools/layoutValidation/types';
import type { IFormData } from 'src/features/formData';
import type { ComponentValidation, GroupValidation, PropsFromGenericComponent } from 'src/layout';
import type { CompExternalExact, CompInternal, HierarchyDataSources } from 'src/layout/layout';
import type { SummaryRendererProps } from 'src/layout/LayoutComponent';
import type { ComponentHierarchyGenerator } from 'src/utils/layout/HierarchyGenerator';
import type { LayoutNode } from 'src/utils/layout/LayoutNode';
import type { LayoutPage } from 'src/utils/layout/LayoutPage';
import type { IValidationContext, IValidationObject, ValidationContextGenerator } from 'src/utils/validation/types';

export class Group extends GroupDef implements GroupValidation, ComponentValidation {
  private _hierarchyGenerator = new GroupHierarchyGenerator();

  directRender(): boolean {
    return true;
  }

  render(props: PropsFromGenericComponent<'Group'>): JSX.Element | null {
    return <GroupRenderer {...props} />;
  }

  renderSummary({
    onChangeClick,
    changeText,
    summaryNode,
    targetNode,
    overrides,
  }: SummaryRendererProps<'Group'>): JSX.Element | null {
    return (
      <SummaryGroupComponent
        onChangeClick={onChangeClick}
        changeText={changeText}
        summaryNode={summaryNode}
        targetNode={targetNode}
        overrides={overrides}
      />
    );
  }

  renderSummaryBoilerplate(): boolean {
    return false;
  }

  getDisplayData(): string {
    return '';
  }

  hierarchyGenerator(): ComponentHierarchyGenerator<'Group'> {
    return this._hierarchyGenerator;
  }

  runComponentValidation(
    node: LayoutNode<'Group'>,
    { langTools }: IValidationContext,
    _overrideFormData?: IFormData,
  ): IValidationObject[] {
    if (!node.isRepGroup()) {
      return [];
    }

    const validationObjects: IValidationObject[] = [];
    // check if minCount is less than visible rows
    const repeatingGroupComponent = node.item;
    const repeatingGroupMinCount = repeatingGroupComponent.minCount || 0;
    const repeatingGroupVisibleRows = repeatingGroupComponent.rows.filter(
      (row) => row && !row.groupExpressions?.hiddenRow,
    ).length;

    const repeatingGroupMinCountValid = repeatingGroupMinCount <= repeatingGroupVisibleRows;

    // if not valid, return appropriate error message
    if (!repeatingGroupMinCountValid) {
      const errorMessage = langTools.langAsString('validation_errors.minItems', [repeatingGroupMinCount]);

      validationObjects.push(buildValidationObject(node, 'errors', errorMessage, 'group'));
    }

    return validationObjects;
  }

  runGroupValidations(
    node: LayoutNode<'Group'>,
    validationCtxGenerator: ValidationContextGenerator,
    onlyInRowIndex?: number,
  ): IValidationObject[] {
    return runValidationOnNodes(node.flat(true, onlyInRowIndex), validationCtxGenerator);
  }

  makeNode(
    item: CompInternal<'Group'>,
    parent: LayoutNode | LayoutPage,
    top: LayoutPage,
    dataSources: HierarchyDataSources,
    rowIndex?: number,
  ): LayoutNodeForGroup {
    return new LayoutNodeForGroup(item, parent, top, dataSources, rowIndex);
  }

  isDataModelBindingsRequired(node: LayoutNode<'Group'>): boolean {
    return node.isRepGroup() || node.isRepGroupLikert();
  }

  validateDataModelBindings(ctx: LayoutValidationCtx<'Group'>): string[] {
    const [errors, result] = this.validateDataModelBindingsAny(ctx, 'group', ['array']);
    if (errors) {
      return errors;
    }

    if (result) {
      const innerType = Array.isArray(result.items) ? result.items[0] : result.items;
      if (!innerType || typeof innerType !== 'object' || !innerType.type || innerType.type !== 'object') {
        return [`group-datamodellbindingen peker mot en ukjent type i datamodellen`];
      }
    }

    return [];
  }

  /**
   * Override layout validation to select a specific pointer depending on the type of group.
   */
  validateLayoutConfing(
    component: CompExternalExact<'Group'>,
    validatate: (pointer: string, data: unknown) => ErrorObject[] | undefined,
  ): ErrorObject[] | undefined {
    let schemaPointer = '#/definitions/AnyComponent';
    if (groupIsNonRepeatingExt(component)) {
      schemaPointer = '#/definitions/CompGroupNonRepeating';
    } else if (groupIsNonRepeatingPanelExt(component)) {
      schemaPointer = '#/definitions/CompGroupNonRepeatingPanel';
    } else if (groupIsRepeatingLikertExt(component)) {
      schemaPointer = '#/definitions/CompGroupRepeatingLikert';
    } else if (groupIsRepeatingExt(component)) {
      schemaPointer = '#/definitions/CompGroupRepeating';
    }
    return validatate(schemaPointer, component);
  }
}
