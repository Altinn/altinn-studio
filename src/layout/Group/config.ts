import { CG, Variant } from 'src/codegen/CG';
import { CompCategory } from 'src/layout/common';

export const Config = new CG.component({
  category: CompCategory.Container,
  rendersWithLabel: false,
  capabilities: {
    renderInTable: false,
    renderInButtonGroup: false,
    renderInAccordion: false,
    renderInAccordionGroup: false,
    renderInCards: true,
    renderInCardsMedia: false,
  },
})
  .addTextResource(
    new CG.trb({
      name: 'title',
      title: 'Title',
      description: 'The title of the group (shown above the group)',
    }),
  )
  .addTextResource(
    new CG.trb({
      name: 'description',
      title: 'Description',
      description: 'The description text shown underneath the title',
    }),
  )
  .addProperty(new CG.prop('childComponents', new CG.arr(CG.layoutNode)).onlyIn(Variant.Internal))
  .addProperty(
    new CG.prop(
      'groupingIndicator',
      new CG.enum('indented', 'panel')
        .optional()
        .setTitle('Set grouping indicator')
        .setDescription('Can visually group components together by indenting them or by putting them in a panel. '),
    ),
  )
  .addProperty(
    new CG.prop(
      'children',
      new CG.arr(new CG.str())
        .setTitle('Children')
        .setDescription('Array of component IDs that should be displayed in the group'),
    ).onlyIn(Variant.External),
  )
  .addProperty(
    new CG.prop(
      'headingLevel',
      new CG.enum(2, 3, 4, 5, 6)
        .optional()
        .setTitle('Heading level')
        .setDescription('The heading level of the group title.'),
    ),
  );
