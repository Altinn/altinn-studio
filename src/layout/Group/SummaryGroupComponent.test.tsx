import React from 'react';

import configureStore from 'redux-mock-store';

import { getFormDataStateMock } from 'src/__mocks__/formDataStateMock';
import { getFormLayoutStateMock } from 'src/__mocks__/formLayoutStateMock';
import { getInitialStateMock } from 'src/__mocks__/initialStateMock';
import { SummaryGroupComponent } from 'src/layout/Group/SummaryGroupComponent';
import { renderWithProviders } from 'src/test/renderWithProviders';
import { useResolvedNode } from 'src/utils/layout/ExprContext';
import type { LayoutNode } from 'src/utils/layout/LayoutNode';

describe('SummaryGroupComponent', () => {
  let mockHandleDataChange: () => void;
  let mockStore;

  beforeAll(() => {
    const createStore = configureStore();
    const formLayout = getFormLayoutStateMock({
      layouts: {
        page1: [
          {
            type: 'Group',
            id: 'groupComponent',
            dataModelBindings: {
              group: 'mockGroup',
            },
            textResourceBindings: {
              title: 'mockGroupTitle',
            },
            children: ['0:mockId1', '1:mockId2'],
            edit: {
              multiPage: true,
            },
            maxCount: 3,
          },
          {
            type: 'Input',
            id: 'mockId1',
            dataModelBindings: {
              simpleBinding: 'mockGroup.mockDataBinding1',
            },
            readOnly: false,
            required: false,
            textResourceBindings: {
              title: 'mockField1',
            },
            triggers: [],
          },
          {
            type: 'Input',
            id: 'mockId2',
            dataModelBindings: {
              simpleBinding: 'mockGroup.mockDataBinding2',
            },
            readOnly: false,
            required: false,
            textResourceBindings: {
              title: 'mockField2',
            },
            triggers: [],
          },
          {
            type: 'Summary',
            id: 'mySummary',
            componentRef: 'groupComponent',
            largeGroup: false,
          },
        ],
      },
      uiConfig: {
        focus: null,
        hiddenFields: [],
        repeatingGroups: {
          groupComponent: {
            index: 0,
            dataModelBinding: 'mockGroup',
          },
        },
        currentView: 'page1',
        navigationConfig: {},
        tracks: {
          order: ['page1'],
          hidden: [],
          hiddenExpr: {},
        },
        excludePageFromPdf: [],
        excludeComponentFromPdf: [],
      },
    });

    const formData = getFormDataStateMock({
      formData: {
        'mockGroup[0].mockDataBinding1': '1',
        'mockGroup[0].mockDataBinding2': '2',
      },
    });

    const initialState: any = getInitialStateMock({
      formData,
      formLayout,
      textResources: {
        error: null,
        language: 'nb',
        resourceMap: {
          mockGroupTitle: {
            value: 'Mock group',
          },
          mockField1: {
            value: 'Mock field 1',
          },
          mockField2: {
            value: 'Mock field 2',
          },
        },
      },
    });
    mockStore = createStore(initialState);
  });

  beforeEach(() => {
    mockHandleDataChange = jest.fn();
  });

  test('SummaryGroupComponent -- should match snapshot', () => {
    const { asFragment } = renderSummaryGroupComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  function renderSummaryGroupComponent() {
    function Wrapper() {
      const summaryNode = useResolvedNode('mySummary') as LayoutNode<'Summary'>;
      const groupNode = useResolvedNode('groupComponent') as LayoutNode<'Group'>;

      return (
        <SummaryGroupComponent
          changeText={'Change'}
          onChangeClick={mockHandleDataChange}
          summaryNode={summaryNode}
          targetNode={groupNode}
        />
      );
    }

    return renderWithProviders(<Wrapper />, { store: mockStore });
  }
});
