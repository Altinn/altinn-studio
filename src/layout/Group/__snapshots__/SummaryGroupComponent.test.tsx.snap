// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryGroupComponent SummaryGroupComponent -- should match snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="summary-group-component"
    style="width: 100%;"
  >
    <div
      class="container"
    >
      <span
        class="label"
      >
        <span>
          Mock group
        </span>
      </span>
      <button
        aria-label="Change: Mock group"
        class="Button-module_button__oVUcI utility-module_focusable__y1VLB Button-module_small__X1q5F Button-module_tertiary__eJ-qG Button-module_second__h-x1a"
        type="button"
      >
        Change
        <svg
          aria-hidden="true"
          class="Button-module_icon__Fb-nR"
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M22.835 1.165a3.976 3.976 0 0 1 0 5.623L8.073 21.549.682 24 0 23.318l2.45-7.392L17.21 1.165a3.977 3.977 0 0 1 5.624 0Zm-4.218 7.029-2.811-2.812L4.188 17l-1.393 4.205 4.207-1.395L18.618 8.194ZM21.43 2.57a1.989 1.989 0 0 0-2.703-.1l-.108.1-1.406 1.406 2.811 2.812 1.406-1.406a1.988 1.988 0 0 0 .101-2.703l-.1-.109Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
    <div
      style="width: 100%;"
    >
      <div
        class="border"
      >
        <div
          data-testid="summary-item-compact"
        >
          <span>
            <span>
              Mock field 1
            </span>
             : 
          </span>
          <span
            class="data"
          >
            1
          </span>
        </div>
        <div
          data-testid="summary-item-compact"
        >
          <span>
            <span>
              Mock field 2
            </span>
             : 
          </span>
          <span
            class="data"
          >
            2
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
